test/tools/ossfuzz/SolCustomMutator.cpp

```
@@ -0,0 +1,87 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <test/tools/ossfuzz/SolCustomMutator.h>
+
+#include <antlr4-runtime.h>
+#include <cstring>
+
+using namespace std;
+using namespace antlr4;
+using namespace solidity::test::fuzzer;
+
+namespace {
+/// Forward declare libFuzzer's default mutator definition
+extern "C" size_t LLVMFuzzerMutate(uint8_t* _data, size_t _size, size_t _maxSize);
+
+/// Define Solidity's custom mutator by implementing libFuzzer's
+/// custom mutator external interface.
+extern "C" size_t LLVMFuzzerCustomMutator(
+	uint8_t* _data,
+	size_t _size,
+	size_t _maxSize,
+	unsigned int _seed
+)
+{
+	if (_maxSize <= _size || _size == 0)
+		return LLVMFuzzerMutate(_data, _size, _maxSize);
+	return SolCustomMutator{_data, _size, _maxSize, _seed}.mutate();
+}
+}
+
+SolCustomMutator::SolCustomMutator(uint8_t* _data, size_t _size, size_t _maxSize, unsigned int _seed)
+{
+	Data = _data;
+	Size = _size;
+	In = string(_data, _data + _size),
+	MaxMutantSize = _maxSize;
+	Seed = _seed;
+}
+
+size_t SolCustomMutator::mutate()
+{
+	try
+	{
+		antlr4::ANTLRInputStream AStream(In);
+		SolidityLexer Lexer(&AStream);
+		Lexer.removeErrorListeners();
+		antlr4::CommonTokenStream Tokens(&Lexer);
+		Tokens.fill();
+		SolidityParser Parser(&Tokens);
+		Parser.removeErrorListeners();
+		SolidityMutator Mutator(Seed);
+		Parser.sourceUnit()->accept(&Mutator);
+		Out = Mutator.toString();
+
+		// If mutant is empty, default to libFuzzer's mutator
+		if (Out.empty())
+			return LLVMFuzzerMutate(Data, Size, MaxMutantSize);
+		else
+		{
+			size_t mutantSize = Out.size() >= MaxMutantSize ? MaxMutantSize - 1 : Out.size();
+			mempcpy(Data, Out.data(), mutantSize);
+			return mutantSize;
+		}
+	}
+	// Range error is thrown by antlr4 runtime's ANTLRInputStream
+	// See https://github.com/antlr/antlr4/issues/2036
+	catch (range_error const&)
+	{
+		// Default to libFuzzer's mutator
+		return LLVMFuzzerMutate(Data, Size, MaxMutantSize);
+	}
```

I'd put the handler over the smallest scope possible, not to silence this exception being thrown from other places for unrelated reasons. I.e. only around `antlr4::ANTLRInputStream AStream(In);`.